File: Algorithm_349_Intersection of Two Arrays.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
vector<int> intersection(vector<int>& nums1, vector<int>& nums2)
{
	sort(nums1.begin(),nums1.end());
	sort(nums2.begin(),nums2.end());
	int p1=0,p2=0,len1=nums1.size(),len2=nums2.size();
	vector<int> res;
	if (len1==0 || len2==0) return res;
	while (p1<len1 && p2<len2)
	{
		if (nums1[p1]==nums2[p2] && (res.size()==0 || nums1[p1]!=res.back()))
		{
			res.push_back(nums1[p1]);
			p1++;
			p2++;
		}
		else if (nums1[p1]>nums2[p2]) p2++;
		else p1++;
	}
	return res;
}
int main()
{
	vector<int> n1, n2;
	n1.push_back(1);
	n2.push_back(1);
	n2.push_back(1);
	vector<int> res=intersection(n1,n2);
	for (int i=0;i<res.size();i++) cout<<res[i]<<' ';
	return 0;
}

File: Algorithm_349_Intersection of Two Arrays.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;

// 方法1：把nums1中所有数字存入set，然后遍历nums2，查当前数字是否存在set中
// 若存在，放入result，result也是一个hashset，用于去重
// 最后，把result中的数字存入int[]并返回

class Solution {
    public int[] intersection(int[] nums1, int[] nums2) {
        Set<Integer> set = new HashSet<>();
        for (int num : nums1) {
            set.add(num);
        }
        Set<Integer> result = new HashSet<>();
        for (int num : nums2) {
            if (set.contains(num)) {
                result.add(num);
            }
        }
        int[] resArr = new int[result.size()];
        int idx = 0;
        for (int num : result) {
            resArr[idx++] = num;
        }
        return resArr;
    }
}

// stream写法
class StreamSolution {
    public int[] intersection(int[] nums1, int[] nums2) {
        Set<Integer> set = Arrays.stream(nums1).boxed().collect(Collectors.toSet());
        return Arrays.stream(nums2).distinct().filter(set::contains).toArray();
    }
}

// 方法2：排序+双指针

class Solution {
    public int[] intersection(int[] nums1, int[] nums2) {
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        int length1 = nums1.length, length2 = nums2.length;
        int[] intersection = new int[length1 + length2];
        int index = 0, index1 = 0, index2 = 0;
        while (index1 < length1 && index2 < length2) {
            int num1 = nums1[index1], num2 = nums2[index2];
            if (num1 == num2) {
                // 保证加入元素的唯一性
                if (index == 0 || num1 != intersection[index - 1]) {
                    intersection[index++] = num1;
                }
                index1++;
                index2++;
            } else if (num1 < num2) {
                index1++;
            } else {
                index2++;
            }
        }
        return Arrays.copyOfRange(intersection, 0, index);
    }
}